### ISSUE TITLE
Sequence classification with LSTM + max pooling

### ISSUE THREAD

**Issue by rfeinman** [fchollet/keras#1459]
Hello,

I am trying to build an LSTM for sequence classification. I would like the network to be able to accept variable length inputs. Here is an example of what the inputs would look like:

inputs after applying pad_sequences:
X = [[0, 0, 0, 1, 2], [0, 1, 4, 3, 5], [0, 0, 1, 1, 5]]
binary class labels:
y = [1, 0, 1]

I am looking for something as defined below. This model uses a 10-word vocabulary, an embedding that converts words into 5-element vectors, and a single 5-node hidden LSTM layer that returns sequences. A pooling layer accepts the temporal sequence output by the LSTM layer and performs temporal max-pooling, looking at only the non-masked portion of the sequence. The pooling layer converts the entire variable-length hidden vector sequence into a single hidden vector, and then feeds its output to the Dense layer.

model = Sequential()
model.add(Embedding(input_dim=10, output_dim=5, mask_zero=True))
model.add(LSTM(output_dim=5, activation='sigmoid', return_sequences=True))
model.add(MaxPool()) ## <----  This is what I need
model.add(Dense(output_dim=1, activation='sigmoid'))
model.compile(loss='binary_crossentropy', optimizer='rmsprop', class_mode='binary')

model.fit(X, y)

Any thoughts on how to best facilitate this?

Thanks,
Reuben

**Comment by rfeinman** [comment]
This did the trick:

from keras.layers.core import MaskedLayer

class MaxPool(MaskedLayer):
    def __init__(self):
        super(MaskedLayer, self).__init__()
        self.input = T.tensor3()

    @property
    def output_shape(self):
        # remove temporal dimension
        return (self.input_shape[0], self.input_shape[2])

    def get_output_mask(self, train=False):
        return None

    def get_output(self, train=False):
        data = self.get_input(train)
        mask = self.get_input_mask(train)
        if mask is None:
            mask = T.sum(T.ones_like(data), axis=-1)
        mask = mask.dimshuffle(0, 1, "x")
        masked_data = T.switch(T.eq(mask, 0), -np.inf, data)

        return masked_data.max(axis=1)

**Comment by rfeinman** [comment]
This did the trick!

class MaxPool(MaskedLayer):
    def __init__(self):
        super(MaskedLayer, self).__init__()
        self.input = T.tensor3()

    @property
    def output_shape(self):
        # remove temporal dimension
        return (self.input_shape[0], self.input_shape[2])

    def get_output_mask(self, train=False):
        return None

    def get_output(self, train=False):
        data = self.get_input(train)
        mask = self.get_input_mask(train)
        if mask is None:
            mask = T.sum(T.ones_like(data), axis=-1)
        mask = mask.dimshuffle(0, 1, "x")
        masked_data = T.switch(T.eq(mask, 0), -np.inf, data)

        return masked_data.max(axis=1)

**Comment by rfeinman** [comment]
class MaxPooling(MaskedLayer):
    def __init__(self):
        super(MaskedLayer, self).__init__()
        self.input = T.tensor3()

    def output_shape(self):
        # remove temporal dimension
        return (self.input_shape[0], self.input_shape[2])

    def get_output_mask(self, train=False):
        return None

    def get_output(self, train=False):
        data = self.get_input(train)
        mask = self.get_input_mask(train)
        if mask is None:
            mask = T.sum(T.ones_like(data), axis=-1)
        mask = mask.dimshuffle(0, 1, "x")
        masked_data = T.switch(T.eq(mask, 0), -np.inf, data)

        return masked_data.max(axis=1)

**Comment by chenchenbox** [comment]
It performs well. Thanks.
